compio-runtime/src/runtime/mod.rs

```diff
@@ -187,8 +195,259 @@ impl Runtime {
+    /// panic and possible memory leak.
+    ///
+    /// ```should_panic
+    /// use compio_runtime::Runtime;
+    ///
+    /// let rt1 = Runtime::new().unwrap();
+    /// let rt2 = Runtime::new().unwrap();
+    ///
+    /// let enter1 = rt1.enter();
+    /// let enter2 = rt2.enter();
+    ///
+    /// drop(enter1);
+    /// drop(enter2);
+    /// ```
+    pub fn enter(&self) -> EnterGuard {
+        EnterGuard::new(self)
+    }
+
+    /// Start a compio runtime and block on the future till it completes.
+    pub fn block_on<F: Future>(&self, future: F) -> F::Output {
+        let _guard = self.enter();
+        self.inner.block_on(future)
+    }
+
+    /// Spawns a new asynchronous task, returning a [`Task`] for it.
+    ///
+    /// Spawning a task enables the task to execute concurrently to other tasks.
+    /// There is no guarantee that a spawned task will execute to completion.
+    pub fn spawn<F: Future + 'static>(&self, future: F) -> Task<F::Output> {
+        self.inner.spawn(future)
+    }
+
+    /// Attach a raw file descriptor/handle/socket to the runtime.
+    ///
+    /// You only need this when authoring your own high-level APIs. High-level
+    /// resources in this crate are attached automatically.
+    pub fn attach(&self, fd: RawFd) -> io::Result<()> {
+        self.inner.attach(fd)
+    }
+
+    /// Submit an operation to the runtime.
+    ///
+    /// You only need this when authoring your own [`OpCode`].
+    pub fn submit<T: OpCode + 'static>(&self, op: T) -> impl Future<Output = BufResult<usize, T>> {
+        self.inner.submit(op)
+    }
+}
+
+impl AsRawFd for Runtime {
+    fn as_raw_fd(&self) -> RawFd {
+        self.inner.as_raw_fd()
+    }
+}
+
+#[cfg(feature = "criterion")]
+impl criterion::async_executor::AsyncExecutor for Runtime {
+    fn block_on<T>(&self, future: impl Future<Output = T>) -> T {
+        self.block_on(future)
+    }
+}
+
+#[cfg(feature = "criterion")]
+impl criterion::async_executor::AsyncExecutor for &Runtime {
+    fn block_on<T>(&self, future: impl Future<Output = T>) -> T {
+        (**self).block_on(future)
+    }
+}
+
+/// Builder for [`Runtime`].
+#[derive(Debug, Clone)]
+pub struct RuntimeBuilder {
+    proactor_builder: ProactorBuilder,
+}
+
+impl Default for RuntimeBuilder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl RuntimeBuilder {
+    /// Create the builder with default config.
+    pub fn new() -> Self {
+        Self {
+            proactor_builder: ProactorBuilder::new(),
+        }
+    }
+
+    /// Replace proactor builder.
+    pub fn with_proactor(&mut self, builder: ProactorBuilder) -> &mut Self {
+        self.proactor_builder = builder;
+        self
+    }
+
+    /// Build [`Runtime`].
+    pub fn build(&self) -> io::Result<Runtime> {
+        Ok(Runtime {
+            inner: Rc::new(RuntimeInner::new(&self.proactor_builder)?),
+        })
+    }
+}
+
+/// Runtime context guard, exists the runtime context on drop.
```

Comment:
Runtime context guard.

When the guard is dropped, exit the corresponding runtime context.